server/src/main/java/org/opensearch/common/settings/Setting.java

```diff
@@ -2337,11 +2337,17 @@ public static <T> Setting<List<T>> listSetting(
         if (defaultStringValue.apply(Settings.EMPTY) == null) {
             throw new IllegalArgumentException("default value function must not return null");
         }
-        Function<String, List<T>> parser = (s) -> parseableStringToList(s).stream().map(singleValueParser).collect(Collectors.toList());
+        Function<String, List<T>> parser = (s) -> isParsedAsEmptyArray(s)
```

Comment:
@machenity this is not how the list setting is supposed to be processed, the `[]` string from YAML file is processed correctly, you could check it by improving the SettingsTests::testSimpleYamlSettings

```
test1:
  value1: value1
  test2:
    value2: value2
    value3: 2
  test3:
    - test3-1
    - test3-2
test4: []
```

```
    public void testSimpleYamlSettings() throws Exception {
        final String yaml = "/org/opensearch/common/settings/loader/test-settings.yml";
        final Settings settings = Settings.builder().loadFromStream(yaml, getClass().getResourceAsStream(yaml), false).build();

        assertThat(settings.get("test1.value1"), equalTo("value1"));
        assertThat(settings.get("test1.test2.value2"), equalTo("value2"));
        assertThat(settings.getAsInt("test1.test2.value3", -1), equalTo(2));

        // check array
        assertNull(settings.get("test1.test3.0"));
        assertNull(settings.get("test1.test3.1"));
        assertThat(settings.getAsList("test1.test3").size(), equalTo(2));
        assertThat(settings.getAsList("test1.test3").get(0), equalTo("test3-1"));
        assertThat(settings.getAsList("test1.test3").get(1), equalTo("test3-2"));
        assertThat(settings.getAsList("test4"), hasSize(0));
    }
```

Reply:
Right. But That's not a point what I want to solve. If we edit the SettingsTests::testSimpleYamlSettings:
```yaml
test1:
  value1: value1
  test2:
    value2: value2
    value3: 2
  test3:
    - test3-1
    - test3-2
  test4: []
  test5: ${NODE_ROLE}
```

```java
public void testSimpleYamlSettings() throws Exception {
        final String yaml = "/org/opensearch/common/settings/loader/test-settings.yml";
        final Settings settings = Settings.builder().loadFromStream(yaml, getClass().getResourceAsStream(yaml), false)
            .replacePropertyPlaceholders() // Replace placeholders with env vars
            .build();

       // ...

        // check array
        assertThat(settings.getAsList("test1.test4"), empty()); // this succeeded
        assertThat(settings.getAsList("test1.test5"), empty()); // this failed
    }
```
If we run the test above with an environment variable `NODE_ROLE=[]`, you can see the last assertion failed.
<img width="705" alt="image" src="https://github.com/opensearch-project/OpenSearch/assets/16838736/d621150e-b91c-4e90-b6a7-a7473880989d">

<img width="648" alt="image" src="https://github.com/opensearch-project/OpenSearch/assets/16838736/2ac100ec-3e6c-4629-baa5-0ce136988d4f">

Reply:
Got the use case now, thanks @machenity (it was a long tail issue). I think the difficulty here is that we don't know what the setting value should be since the placeholders are applied after parsing the YAML / JSON and the semantic of the value is not known anymore. We could enhance the placeholder substitution to recognize that:

```
                // if the values exists and has length, we should maintain it in the map
                // otherwise, the replace process resolved into removing it
                if ("[]".equalsIgnoreCase(value)) { 
                    entry.setValue(Collections.emptyList());
                } else if (Strings.hasLength(value)) {
                    entry.setValue(value);
                } else  {
                    entryItr.remove();
                }
```

But that would be not sufficient as per https://github.com/opensearch-project/OpenSearch/issues/3412#issuecomment-1134724811 (if we accept `[]`, we should accept `["a", "b", "c"]` as well. However changing the behaviour of treating the setting values is super sensitive.

Reply:
@machenity I think I have drafted an acceptable solution, do you mind if I push the change there or you would prefer me to open a separate pull request? Thank you.

Reply:
> But that would be not sufficient as per https://github.com/opensearch-project/OpenSearch/issues/3412#issuecomment-1134724811 (if we accept [], we should accept ["a", "b", "c"] as well). However changing the behaviour of treating the setting values is super sensitive.

You're right. I want to minimize the impact to the user semantics by treating only [] as a special case, but we should accept ["a", "b", "c"] as well if we accept [], as you told.

@reta Feel free to push here or do other way you prefer. Thanks for your kind reviews and comments.

Reply:
Done, @msfroh may I ask you please to recheck this change? I came up to the same conclusion as you, the change is now isolated to placeholders  (environment variables) only, thanks!

Reply:
Awesome! Thanks for scoping it down 👍